### PR TITLE
[Design] Create Template engine from project snapshot

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

**Comment by rynowak** [comment]
I have not updated all of the tests yet, but just check out how this works now.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTemplateEngineFactoryService.cs [hunk 1]
```diff
@@ -31,17 +36,20 @@ public override RazorTemplateEngine Create(string projectPath, Action<IRazorEngi
...
             }
 
+            // In 15.5 we expect projectPath to be a directory, NOT the path to the csproj.
+            var project = FindProject(projectPath);
+            var configuration = project?.Configuration ?? DefaultConfiguration;
+
             RazorEngine engine;
-            var mvcVersion = GetMvcVersion(projectPath);
-            if (mvcVersion?.Major == 1)
+            if (configuration.RazorAssembly.Identity.Version.Major == 1)
```
Comment (rynowak) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTemplateEngineFactoryService.cs]: We'll make this less hardcoded in the future when we do 3rd party extensibility

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTemplateEngineFactoryService.cs [hunk 2]
```diff
@@ -72,28 +74,20 @@ public override RazorTemplateEngine Create(string projectPath, Action<IRazorEngi
...
-            {
-                var compilation = CSharpCompilation.Create(project.AssemblyName).AddReferences(project.MetadataReferences);
-
-                foreach (var identity in compilation.ReferencedAssemblyNames)
+                var project = projects[i];
+                if (project.UnderlyingProject.FilePath != null)
                 {
-                    if (identity.Name == MvcAssemblyName)
+                    Path.GetDirectoryName(project.UnderlyingProject.FilePath);
+                    if (string.Equals(directory, NormalizeDirectoryPath(Path.GetDirectoryName(project.UnderlyingProject.FilePath)), StringComparison.OrdinalIgnoreCase))
```
Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTemplateEngineFactoryService.cs]: Doing ignore case here might be troublesome once VS for mac pops into the picture.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTemplateEngineFactoryService.cs [hunk 3]
```diff
@@ -72,28 +74,20 @@ public override RazorTemplateEngine Create(string projectPath, Action<IRazorEngi
...
-            if (project != null)
-            {
-                var compilation = CSharpCompilation.Create(project.AssemblyName).AddReferences(project.MetadataReferences);
-
-                foreach (var identity in compilation.ReferencedAssemblyNames)
+                var project = projects[i];
+                if (project.UnderlyingProject.FilePath != null)
                 {
-                    if (identity.Name == MvcAssemblyName)
+                    Path.GetDirectoryName(project.UnderlyingProject.FilePath);
```
Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTemplateEngineFactoryService.cs]: unused

File: src/Microsoft.VisualStudio.LanguageServices.Razor/Legacy/LegacyTemplateEngineFactoryService.cs [hunk 4]
```diff
@@ -15,24 +17,46 @@ namespace Microsoft.VisualStudio.LanguageServices.Razor
...
-                // For now we're hardcoded to use MVC's extensibility.
-                RazorExtensions.Register(b);
-            });
+            if (inner == null)
+            {
+                throw new ArgumentNullException(nameof(inner));
+            }
+
+            _workspace = workspace;
+            _inner = workspace.Services.GetLanguageServices(RazorLanguage.Name).GetRequiredService<Inner>();
```
Comment (NTaylorMullen) [inline src/Microsoft.VisualStudio.LanguageServices.Razor/Legacy/LegacyTemplateEngineFactoryService.cs]: Use the one from the ctor.

File: src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTemplateEngineFactoryService.cs (hunk 2 above)
Comment (rynowak) [reply src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultTemplateEngineFactoryService.cs]: This whole thing is inexact

File: src/Microsoft.VisualStudio.LanguageServices.Razor/Legacy/LegacyTemplateEngineFactoryService.cs (hunk 4 above)
Comment (rynowak) [reply src/Microsoft.VisualStudio.LanguageServices.Razor/Legacy/LegacyTemplateEngineFactoryService.cs]: 👍

**Comment by rynowak** [comment]
Updated and fixed tests

**Pull request merged**